oap-server/server-library/library-datacarrier-queue/src/main/java/org/apache/skywalking/oap/server/library/datacarrier/buffer/ArrayBlockingQueueBuffer.java

```diff
@@ -40,8 +40,10 @@ public class ArrayBlockingQueueBuffer<T> implements QueueBuffer<T> {
 
     @Override
     public boolean save(T data) {
-        //only BufferStrategy.BLOCKING
         try {
+            if (BufferStrategy.IF_POSSIBLE.equals(strategy) && queue.remainingCapacity() == 0) {
+                return false;
+            }
```

Comment:
From the javadoc of `queue.remainingCapacity`, this is a misuse

```
     * <p>Note that you <em>cannot</em> always tell if an attempt to insert
     * an element will succeed by inspecting {@code remainingCapacity}
     * because it may be the case that another thread is about to
     * insert or remove an element.
```

And I'm curious why don't we remove `ArrayBlockingQueueBuffer.java` and keep `Buffer.java`, I personally think the `Buffer` class is more easy to read.

Reply:
Because the ArrayBlockingQueueBuffer has better performance from x-thread visibility side. We have tested and proved that.

Reply:
For the edge case, I think we don't need to be accurate. Because accurate means more blocks.

Reply:
For the edge cases, (1) if the queue is not full but `queue.remainingCapacity() == 0`, we drop some data that should not have been dropped, this is not a big deal, it doesn't worsen the case, but, (2) if the queue is full but `queue.remainingCapacity() != 0`, `queue.remainingCapacity()` itself needs to acquire a lock and then `queue.put()` also needs the same again, AND `queue.put` is blocking, this worsen the edge case where the queue is already full.

I think we can leverage the `offer` method to reduce acquisition of locks and avoid unnecessary blocking

```suggestion
            if (BufferStrategy.IF_POSSIBLE.equals(strategy)) {
                return queue.offer(data);
            }
```